packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,618 @@
+        TransferBatch[] memory operations,
+        bytes           memory txsData
+        )
+        internal
+    {
+        for (uint o = 0; o < operations.length; o++) {
+            InternalBridgeTransfer[] memory transfers = operations[o].transfers;
+
+            // Check if these transfers can be processed
+            bytes32 hash = _hashTransfers(operations[o].batchID, transfers);
+            require(!_arePendingTransfersTooOld(hash), "TRANSFERS_TOO_OLD");
+
+            // Mark transfers as completed
+            pendingTransfers[hash] = 0;
+
+            // Verify transfers
+            TransferTransaction.Transfer memory transfer;
+            for (uint i = 0; i < transfers.length; i++) {
+                TransferTransaction.readTx(txsData, ctx.txIdx++ * ExchangeData.TX_DATA_AVAILABILITY_SIZE, transfer);
+
+                uint16 tokenID = transfers[i].tokenID;
+                require(
+                    transfer.fromAccountID == ctx.accountID &&
+                    transfer.to == transfers[i].owner &&
+                    transfer.from == address(this) &&
+                    transfer.tokenID == tokenID &&
+                    transfer.feeTokenID == tokenID &&
+                    _isAlmostEqualAmount(transfer.amount, transfers[i].amount) &&
+                    transfer.fee <= (uint(transfer.amount).mul(MAX_FEE_BIPS) / 10000),
+                    "INVALID_BRIDGE_TRANSFER_TX_DATA"
+                );
+            }
+        }
+    }
+
+    function _processCalls(
+        Context          memory ctx,
+        BridgeOperations memory operations,
+        bytes            memory txsData
+        )
+        internal
+    {
+        TokenData[] memory tokens = operations.tokens;
+
+        uint[] memory withdrawalAmounts = new uint[](tokens.length);
+        for (uint i = 0; i < tokens.length; i++) {
+            withdrawalAmounts[i] = tokens[i].amount;
+        }
+
+        // Calls
+        TransferTransaction.Transfer memory transfer;
+        SignatureVerificationTransaction.SignatureVerification memory verification;
+        HashData memory hashData;
+        for (uint c = 0; c < operations.connectorCalls.length; c++) {
+
+            ConnectorCalls memory connectorCall = operations.connectorCalls[c];
+
+            // Verify token data
+            require(connectorCall.tokens.length == tokens.length, "INVALID_TOKEN_DATA");
+            for (uint i = 0; i < tokens.length; i++) {
+                require(tokens[i].token == connectorCall.tokens[i].token, "INVALID_CONNECTOR_TOKEN_DATA");
+                tokens[i].amount = tokens[i].amount.sub(connectorCall.tokens[i].amount);
+            }
+
+            // Call the connector
+            _connectorCall(connectorCall);
+
+            // Verify the transactions
+            for (uint g = 0; g < connectorCall.groups.length; g++) {
+                ConnectorGroup memory group = connectorCall.groups[g];
+                for (uint i = 0; i < group.calls.length; i++) {
```

Comment:
I'm having a hard time understanding the notion of groups.  If we use adding liquidity of two tokens into Uniswap as an example, in such a group, there are two transfers and one Join. So maybe the group should have **multiple transfers** with **one call**. But the current implementation seems to assume each call has one matched transfer.

Reply:
A group is per specific operation. So for a tokenA -> tokenB swap, a group would simply be a list of tokenA transfers with. 
For batch migration a group could also just be a list of all kinds of tokens.

For adding liquidity for tokenA/tokenB the group would just contain 2 bridge calls from the same owner doing the 2 transfers.  But you're right, maybe it does indeed make sense to more tightly encode these together for cases like these with multiple token transfers (similar to how we do things for AMM joints/exits). It could potentially make cases with multiple transfers a little bit cheaper, but I think only worth if it doesn't make the more normal cases more expensive.

Reply:
Ok, lets talk about this later after your optimizations.

Reply:
Kept the bridge call with 1 transfer/bridge call for now to keep things a bit simpler but also the gas cost increased quite a bit the way I implemented it with more arrays that needed to passed on-chain (which mostly contained a single element, but still has some noticeable overhead of ~20% gas cost increase because we're only talking about ~13,000 gas per bridge call, so even ~2,000 gas extra is quite a lot). Most common scenario's will very likely do just a single transfer so I think it should be fine, and multiple bridge calls to do multiple transfers can still easily be submitted together.